File: xylo-microcontroller/lib/kinematics.cpp
#include <math.h>
#include "kinematics.h"
//#include <iostream>
//using namespace std;

const double RADIAN = 0.0174533;

double Kinematics::toRadians(double d){
    return d * RADIAN;
}
double Kinematics::toDegrees(double r){
    return r/RADIAN;
}

double Kinematics::law_of_cosines(double a, double b, double c){
    return toDegrees(acos( ( pow(a,2) + pow(b,2) - pow(c,2) ) / (2*a*b) ) );
}

double Kinematics::tangent_law(double a, double b, double alpha){
    return atan(b*sin(toRadians(-alpha))/(a+b*cos(toRadians(-alpha))));
}

Vector2 Kinematics::getPointOnCircle(Vector2 origin, double radius, double degrees){

    double x = cos(toRadians(degrees)) * radius;
    double y = -sin(toRadians(degrees)) * radius;
    Vector2 vector(x,y);
    vector.add(origin);
    return vector;
}

double Kinematics::getCircleArc(double radius, double x){

    return toDegrees(acos(x/radius));
}

Vector3 Kinematics::getModTop(Vector3 orig, double radius, double degrees){

    Vector2 in(orig.x,orig.z);
    Vector2 out = getPointOnCircle( in ,radius,-90-degrees);
    return {out.x, orig.y, out.y};
}

Vector3 Kinematics::getModSide(Vector3 orig, double radius, double degrees){
    Vector2 in ( orig.z,orig.y);
    Vector2 out = getPointOnCircle( in ,radius,-90+degrees);
    return {orig.x,out.y,out.x};
}

double Kinematics::getDegTop(Vector3 orig, double radius, Vector3 pos){
    return 90-getCircleArc(radius,-(pos.x-orig.x));
}

Vector3 Kinematics::getOffsetModTop(Vector3 effect, double d){
    double dst = Vector2(effect.x,effect.z).dst(Vector2(FLEX_S2.x,FLEX_S2.z));
    double arc_offset = getDegTop(FLEX_S2,dst, effect);
    return getModTop(Vector3(FLEX_S2.x,effect.y,FLEX_S2.z),dst,-d+arc_offset);
}

double Kinematics::getDegSide(Vector3 orig, double radius, Vector3 pos){
    return getCircleArc(radius,pos.z-orig.z);
}

//Forward kinematics

Vector3 Kinematics::getEffectorPos(double servo1, double servo2, double servo3){

    Vector3 flex_s3 = getModSide(FLEX_S2,L2,servo2);
    Vector3 reach = getModSide({STICK_OFFSET,flex_s3.y,flex_s3.z},L3,servo2+servo3);
    return getOffsetModTop(reach,servo1);
}

//Inverse kinematics

Vector3 Kinematics::getDegs(Vector3 pos){

    //d1
    double dst = Vector2(pos.x,pos.z).dst(Vector2(FLEX_S2.x,FLEX_S2.z));
    double arc_offset = getDegTop(FLEX_S2,dst, Vector3(STICK_OFFSET,pos.y,pos.z));
    double d1 = -(getDegTop(FLEX_S2,dst,pos)-arc_offset);

    //d3
    Vector3 effect_norm = getOffsetModTop(pos,-d1);
    double c = Vector3(effect_norm.x-STICK_OFFSET,effect_norm.y,effect_norm.z).dst(FLEX_S2);
    double d3 = 180-law_of_cosines(L3,L2,c);

    //d2
    double d2 = 90 - toDegrees(atan((-(pos.y-FLEX_S2.y))/(-pos.z+FLEX_S2.z)) - tangent_law(L2,L3,d3));
    return {d1,d2,d3};
}

//int main()
//{
//    Kinematics kinematics;
//    kinematics.getEffectorPos(10,15,30).print();
//    return 0;
//}

File: xylo-python/kinematics.py
import math

RADIAN = 0.0174533

FLEX_S2 = [0,1.875,-3.65]
STICK_OFFSET = -0.4
L2 = 1.075
L3 = 1.8

def toRadians(d):
    return d * RADIAN

def toDegrees(d):
    return d/RADIAN

def dist(a,b):
    return math.sqrt(math.pow(b[0]-a[0],2)+math.pow(b[1]-a[1],2))

def dist3(a,b):
    return math.sqrt(math.pow(b[0]-a[0],2)+math.pow(b[1]-a[1],2)+math.pow(b[2]-a[2],2))

def law_of_cosines(a,b,c):
    alpha = math.acos( (math.pow(a,2) + math.pow(b,2) - math.pow(c,2)) / (2*a*b))
    return alpha / RADIAN

def tangent_law(a,b,alpha):
    return math.atan(b * math.sin(toRadians(-alpha))/(a+b*math.cos(toRadians(-alpha))))

def getPointOnCircle(origin, radius, degrees):
    x = math.cos(toRadians(degrees)) * radius + origin[0]
    y = -math.sin(toRadians(degrees)) * radius + origin[1]
    return [x,y]

def getCircleArc(radius, x):
    return (math.acos(x/radius)) / RADIAN

def getModTop(origin, radius, degrees):
    p = [origin[0],origin[2]]
    out = getPointOnCircle(p,radius,-90-degrees)
    return [out[0],origin[1],out[1]]

def getModSide(origin, radius, degrees):
    p = [origin[2], origin[1]]
    out = getPointOnCircle(p, radius, -90+degrees)
    return [origin[0], out[1], out[0]]

def getDegTop(origin, radius, pos):
    return 90-getCircleArc(radius, -(pos[0]-origin[0]))

def getOffsetModTop(effect, d):
    dst = dist([effect[0],effect[2]],[FLEX_S2[0],FLEX_S2[2]])
    arc_off = getDegTop(FLEX_S2,dst,effect)
    p = getModTop([FLEX_S2[0],effect[1],FLEX_S2[2]],dst,-d+arc_off)
    return p

# //Forward kinematics

def getEffectorPos(s):
    flex_s3 = getModSide(FLEX_S2,L2,s[1])
    reach = getModSide([STICK_OFFSET,flex_s3[1],flex_s3[2]],L3,s[1]+s[2])
    return getOffsetModTop(reach,s[0])

# //Inverse kinematics

def getDegs(pos):
    print(pos)
    dst = dist([pos[0],pos[2]],[FLEX_S2[0],FLEX_S2[2]])
    arc_off = getDegTop(FLEX_S2,dst,[STICK_OFFSET,pos[1],pos[2]])
    d1 = -(getDegTop(FLEX_S2,dst,pos)-arc_off)
    effect_norm = getOffsetModTop(pos,-d1)
    c = dist3([effect_norm[0]-STICK_OFFSET,effect_norm[1],effect_norm[2]],FLEX_S2)
    d3 = 180 - law_of_cosines(L3,L2,c)

    d2 = 90 - (math.atan((-(pos[1]-FLEX_S2[1]))/(-pos[2]+FLEX_S2[2])) - tangent_law(L2,L3,d3))/RADIAN
    return [d1,d2,d3]
